llap-server/src/java/org/apache/hadoop/hive/llap/metrics/LlapDaemonExecutorMetrics.java

```diff
@@ -397,4 +425,106 @@ public JvmMetrics getJvmMetrics() {
   public String getName() {
     return name;
   }
+
+  /**
+   * Generate time aware average for data points.
+   * For example if we have 3s when the queue size is 1, and 1s when the queue size is 2 then the
+   * calculated average should be (3*1+1*2)/4 = 1.25.
+   */
+  @VisibleForTesting
+  static class TimedAverageMetrics {
+    private final int windowDataSize;
+    private final long windowTimeSize;
+    private final Data[] data;
+    private int nextPos = 0;
+
+    /**
+     * Creates and initializes the metrics object.
+     * @param windowDataSize The maximum number of samples stored
+     * @param windowTimeSize The time window used to generate the average in nanoseconds
+     */
+    TimedAverageMetrics(int windowDataSize, long windowTimeSize) {
+      this(windowDataSize, windowTimeSize, System.nanoTime());
```

Comment:
Using the current time as default time essentially makes all values valid within the time window. We therefore aggregate just zeros here. Wouldn't it make sense to initialize this with zero to mark all entries "outside of the window" initially?

Reply:
According to the documentation:
"The value returned represents nanoseconds since some fixed but arbitrary origin time (perhaps in the future, so values may be negative)."

That is why I used the System.nanoTime. Will move them outside the window.
Good catch. Thanks!